storage-operator/pkg/controller/volume/volume_controller.go

```diff
@@ -24,149 +26,761 @@ import (
 	return &ReconcileVolume{
-		client: mgr.GetClient(),
-		scheme: mgr.GetScheme(),
-		salt:   salt.NewClient(getAuthCredential(mgr.GetConfig())),
+		client:   mgr.GetClient(),
+		scheme:   mgr.GetScheme(),
+		recorder: mgr.GetRecorder("volume_controller"),
+		salt:     salt.NewClient(getAuthCredential(mgr.GetConfig())),
 	}
 }
 
-// add adds a new Controller to mgr with r as the reconcile.Reconciler
+// Adds a new Controller to `mgr` with `r` as the reconcile.Reconciler.
 func add(mgr manager.Manager, r reconcile.Reconciler) error {
 	// Create a new controller
-	c, err := controller.New("volume-controller", mgr, controller.Options{Reconciler: r})
+	c, err := controller.New(
+		"volume-controller", mgr, controller.Options{Reconciler: r},
+	)
 	if err != nil {
 		return err
 	}
 
 	// Watch for changes to primary resource Volume
-	err = c.Watch(&source.Kind{Type: &storagev1alpha1.Volume{}}, &handler.EnqueueRequestForObject{})
+	err = c.Watch(
+		&source.Kind{Type: &storagev1alpha1.Volume{}},
+		&handler.EnqueueRequestForObject{},
+	)
 	if err != nil {
 		return err
 	}
 
 	// Watch for changes to secondary resource Pods and requeue the owner Volume
-	err = c.Watch(&source.Kind{Type: &corev1.PersistentVolume{}}, &handler.EnqueueRequestForOwner{
-		IsController: true,
-		OwnerType:    &storagev1alpha1.Volume{},
-	})
+	err = c.Watch(
+		&source.Kind{Type: &corev1.PersistentVolume{}},
+		&handler.EnqueueRequestForOwner{
+			IsController: true,
+			OwnerType:    &storagev1alpha1.Volume{},
+		},
+	)
 	if err != nil {
 		return err
 	}
 
 	return nil
 }
 
-// blank assignment to verify that ReconcileVolume implements reconcile.Reconciler
+// Verify that ReconcileVolume implements reconcile.Reconciler.
 var _ reconcile.Reconciler = &ReconcileVolume{}
 
-// ReconcileVolume reconciles a Volume object
+// Reconciles a Volume object.
 type ReconcileVolume struct {
 	// This client, initialized using mgr.Client() above, is a split client
 	// that reads objects from the cache and writes to the apiserver
-	client client.Client
-	scheme *runtime.Scheme
-	salt   *salt.Client
+	client   client.Client
+	scheme   *runtime.Scheme
+	recorder record.EventRecorder
+	salt     *salt.Client
+}
+
+// Get the PersistentVolume associated to the given volume.
+//
+// Return `nil` if no such volume exists.
+func (self *ReconcileVolume) getPersistentVolume(
+	name string,
+) (*corev1.PersistentVolume, error) {
+	pv := &corev1.PersistentVolume{}
+	key := types.NamespacedName{Namespace: "", Name: name}
+
+	if err := self.client.Get(context.TODO(), key, pv); err != nil {
+		if errors.IsNotFound(err) {
+			return nil, nil
+		}
+		return nil, err
+	}
+	return pv, nil
+}
+
+func (self *ReconcileVolume) notifyStateTransition(
+	volume *storagev1alpha1.Volume, oldPhase storagev1alpha1.VolumePhase,
+) {
+	reqLogger := log.WithValues("Request.Name", volume.Name)
+
+	self.recorder.Eventf(
+		volume, corev1.EventTypeNormal, "StateTransition",
+		"volume phase transition from `%s` to `%s`",
```

Comment:
Is putting these strings in backticks 'standard' practice in K8s? The events won't be rendered through ReST/MD or anything, so I'd argue using single ticks `'` would be better.

Reply:
Using backticks is my own habit I think, I'll use single quote instead, no problem.